### DISCUSSION TITLE
Early feedback

### DISCUSSION THREAD

**Discussion by ewels** [ewels/rich-codex#10] [Ideas]
Hi @thijsmie, @willmcgugan,

I added you both as collaborators on this repo in response to https://github.com/ewels/rich-click/discussions/74 so that you can have a poke if you're curious.. (although I know you already rolled your own @willmcgugan!)

There are a few bits missing and it needs a bit of spit and polish still, but any early feedback would be gratefully received 🙏🏻 

This is a late-night hobby project so has been a bit stop-start in its progress.. (Likely to continue to be the case).

Phil

**Comment by thijsmie** [comment]
Hi @ewels!
I've been trying this out and I am really liking it so far! I have a couple of notes so far (I can turn them into issues if you want):
 * Sometimes I don't need all the smartness that rich-codex does for me, I just need a quick image of something. Is this modular enough to also just give me a "stupid" tool that does "run this command and save it here" instead of the whole scanning routine?
 * I have several things that I "screenshot" that are really examples of a scenario. They require running a background script to do some setup that are then discovered and displayed in the terminal. Do you see a way to integrate some kind of init/teardown context for a command?
 * I render TUI's with `rich.live.Live` and Textual which rely on the alternative-screen and will run forever until ctrl-c. It would be cool to be able to "screenshot" these after N seconds and then kill them. I have no idea how feasible that is however.
 * Fiddling with the terminal width "till it looks good" can be very dependend on what command is running. Would it be possible just to set it to a very high number and then let rich-codex crop to a size where it fits exactly?

**Reply by thijsmie** [reply]
Oh, and one more: --use-pty does not set the terminal size of the resulting window. I have looked around and it is a little bit convoluted to do but possible, for example see https://github.com/8vasu/pypty2/blob/master/pty2.py

**Reply by ewels** [reply]
Great! 🙌🏻 Maybe would be better as issues for further discussion, but I can do some quick responses..

> Sometimes I don't need all the smartness that rich-codex does for me, I just need a quick image of something. Is this modular enough to also just give me a "stupid" tool that does "run this command and save it here" instead of the whole scanning routine?

I'm not totally clear what you're after here, but yes I think it can do both. Either just running `rich-codex` to find stuff in the markdown or `rich-codex --command xxx --img-paths yyy` on the command line (or with comparable options in the GitHub Action / the soon-to-exist config file).

> I have several things that I "screenshot" that are really examples of a scenario. They require running a background script to do some setup that are then discovered and displayed in the terminal. Do you see a way to integrate some kind of init/teardown context for a command?

Yeah, see the GitHub Action that I resurrected last night. You should be able to easily to this to prepare any tools / environments that you need and then run the rich-codex when you're ready (see [`basic_example.yml`](https://github.com/ewels/rich-codex/blob/main/.github/workflows/basic_example.yml)).

> I render TUI's with `rich.live.Live` and Textual which rely on the alternative-screen and will run forever until ctrl-c. It would be cool to be able to "screenshot" these after N seconds and then kill them. I have no idea how feasible that is however.

I saw that @willmcgugan did this with his MkDocs version of the same tool. I don't see why not, but it's a lower priority than some of the other stuff right now. If you could dump this into an issue to track that'd be fab 👍🏻  (feel free to create a PR if you know how to approach it).

> Fiddling with the terminal width "till it looks good" can be very dependend on what command is running. Would it be possible just to set it to a very high number and then let rich-codex crop to a size where it fits exactly?

I played for quite a long time to try to set the terminal width of the command and found it pretty much impossible (without altering the source code of the running tool, anyway). However, I hadn't considered capturing the output and then trimming back the width to that automatically. Nice idea.. 🤔 If you could dump this into an issue as well then I can take a look 👍🏻 

> Oh, and one more: `--use-pty` does not set the terminal size of the resulting window. I have looked around and it is a little bit convoluted to do but possible, for example see https://github.com/8vasu/pypty2/blob/master/pty2.py

Yeah this is quite a new addition, as the result of me trying to get coloured output without needing to edit the source code of the running tool to force colour output. I'm a bit out of my depth with it though. Any help would be welcome 🙏🏻

**Reply by thijsmie** [reply]
Thanks for the responses! I am winding down for today, I'll split out into issues and see about contributing the pty thing. For the ctrl-c/capture tool approach I made some hidden options that limit duration and print the final UI state exactly once, so also low priority for me.

**Reply by thijsmie** [reply]
Oh and RE the other automation: we are not using GitHub actions for CI but Azure Pipelines, and anyway the init/teardown is not the same for every command that needs running, so that method will not work for me.

**Reply by ewels** [reply]
Excellent, straight in with the edge cases that I hadn't considered! 😅 You should still be able to set up some kind of CI configuration that runs rich-click multiple times for specific files / commands etc, with custom setup and teardown.

Messy and difficult to maintain, but better than nothing. And the best I can think of sorry.

**Reply by thijsmie** [reply]
Made some issues! One more thing: `import pty` will actually fail on Windows. I don't know if you are targeting windows, but your `setup.cfg` claims OS::Independent 😉.

**Reply by ewels** [reply]
ok good to know. I'm not too fussed as it's not the default, but we should maybe capture the error and fall back with a nice log message. I don't have access to any windows machines though, so could be tricky.

Which leads me to... should really write some tests 🙄

**Comment by thijsmie** [comment]
Hi @ewels!

I've noticed that the order commands are run in is random, due to the usage of a `set()` to store the commands in. Don't you think the order as they appear in a file would be nice to preserve?

**Reply by ewels** [reply]
I guess.. Or alphabetical? That would be easier (no need to track origin, which can be messy as it collapses identical commands so they could come from multiple files).

**Reply by ewels** [reply]
Ok, they're now sorted alphabetically and also show their origin. If the same command is called from multiple sources it'll just show the first one, whichever that happens to be.
